src/theory/strings/regexp_elim.cpp

```diff
@@ -0,0 +1,441 @@
+  {
+    Node c = children[i];
+    Trace("re-elim-debug") << "  " << c << std::endl;
+    success = false;
+    if (c.getKind() == STRING_TO_REGEXP)
+    {
+      success = true;
+      sep_children.push_back(c[0]);
+      // the next gap is initially strict zero
+      gap_minsize.push_back(0);
+      gap_exact.push_back(true);
+    }
+    else if (c.getKind() == REGEXP_STAR && c[0].getKind() == REGEXP_SIGMA)
+    {
+      // found a gap of any size
+      success = true;
+      gap_exact[gap_exact.size() - 1] = false;
+    }
+    else if (c.getKind() == REGEXP_SIGMA)
+    {
+      // add one to the minimum size of the gap
+      success = true;
+      gap_minsize[gap_minsize.size() - 1]++;
+    }
+    if (!success)
+    {
+      Trace("re-elim-debug") << "...cannot handle " << c << std::endl;
+      break;
+    }
+  }
+  // we should always rewrite concatenations that are purely re.allchar
+  // and re.*( re.allchar ).
+  Assert(!success || !sep_children.empty());
+  if (success && !sep_children.empty())
+  {
+    std::vector<Node> conj;
+    // The following constructs a set of constraints that encodes that a
+    // set of string terms are found, in order, in string x.
+    // prev_end stores the current (symbolic) index in x that we are
+    // searching.
+    Node prev_end = d_zero;
+    unsigned gap_minsize_end = gap_minsize.back();
+    bool gap_exact_end = gap_exact.back();
+    std::vector<Node> non_greedy_find_vars;
+    for (unsigned i = 0, size = sep_children.size(); i < size; i++)
+    {
+      Node sc = sep_children[i];
+      if (gap_minsize[i] > 0)
+      {
+        // the gap to this child is at least gap_minsize[i]
+        prev_end =
+            nm->mkNode(PLUS, prev_end, nm->mkConst(Rational(gap_minsize[i])));
+      }
+      Node lensc = nm->mkNode(STRING_LENGTH, sc);
+      if (gap_exact[i])
+      {
+        // if the gap is exact, it is a substring constraint
+        Node curr = prev_end;
+        Node ss = nm->mkNode(STRING_SUBSTR, x, curr, lensc);
+        conj.push_back(ss.eqNode(sc));
+        prev_end = nm->mkNode(PLUS, curr, lensc);
+      }
+      else
+      {
+        // otherwise, we can use indexof to represent some next occurrence
+        if (gap_exact[i + 1] && i + 1 != size)
+        {
+          if (!options::regExpElimAgg())
+          {
+            success = false;
+            break;
+          }
+          // if the gap after this one is strict, we need a non-greedy find
+          // thus, we add a symbolic constant
+          Node k = nm->mkBoundVar(nm->integerType());
+          non_greedy_find_vars.push_back(k);
+          prev_end = nm->mkNode(PLUS, prev_end, k);
+        }
+        Node curr = nm->mkNode(STRING_STRIDOF, x, sc, prev_end);
+        Node idofFind = curr.eqNode(d_neg_one).negate();
+        conj.push_back(idofFind);
+        prev_end = nm->mkNode(PLUS, curr, lensc);
+      }
+      // if applicable, process the last gap
+      if (i == (size - 1))
```

Comment:
Suggestion: could move this outside of the loop and have the loop only go up to `sep_children.size() - 1`.

Reply:
This will be refactored.